module/VuFind/src/VuFind/UrlShortener/Database.php

```diff
@@ -150,17 +176,40 @@ protected function saveAndShortenHash($path, $hash, $length)
             );
         }
         $shorthash = str_pad(substr($hash, 0, $length), $length, '_');
-        $results = $this->table->select(['hash' => $shorthash]);
+
+        $queryBuilder = $this->entityManager->createQueryBuilder();
```

Comment:
I think all of this could be refactored to a getForHash method in the database service. You might be able to add a flag parameter to control whether or not new rows are automatically created to enable sharing between this method and the resolve method.